src/coredump/stacktrace.c

```diff
@@ -399,15 +403,124 @@ static int parse_core(int fd, const char *executable, char **ret, JsonVariant **
 
         free(buf);
 
+        if (r == -EINVAL)
+                log_warning("Failed to generate stack trace: %s", dwfl_errmsg(dwfl_errno()));
+        else if (r < 0)
+                log_warning_errno(r, "Failed to generate stack trace: %m");
+
         return r;
 }
 
-void coredump_parse_core(int fd, const char *executable, char **ret, JsonVariant **ret_package_metadata) {
+int parse_elf_object(int fd, const char *executable, char **ret, JsonVariant **ret_package_metadata) {
+        _cleanup_close_pair_ int error_pipe[2] = { -1, -1 }, return_pipe[2] = { -1, -1 }, json_pipe[2] = { -1, -1 };
         int r;
 
-        r = parse_core(fd, executable, ret, ret_package_metadata);
-        if (r == -EINVAL)
-                log_warning("Failed to generate stack trace: %s", dwfl_errmsg(dwfl_errno()));
-        else if (r < 0)
-                log_warning_errno(r, "Failed to generate stack trace: %m");
+        r = RET_NERRNO(pipe2(error_pipe, O_CLOEXEC));
+        if (r < 0)
+                return r;
+
+        if (ret) {
+                r = RET_NERRNO(pipe2(return_pipe, O_CLOEXEC));
+                if (r < 0)
+                        return r;
+        }
+
+        if (ret_package_metadata) {
+                r = RET_NERRNO(pipe2(json_pipe, O_CLOEXEC));
+                if (r < 0)
+                        return r;
+        }
+
+        /* Parsing possibly malformed data is crash-happy, so fork and avoid looping. */
+        r = safe_fork_full("(sd-parse-elf)",
```

Comment:
I must say I am not entirely convinced about doing this out of process here a second time. I mean, one of the original reasons why the core_pattern handler doesn't process the coredumps itself but instead passes this off to a set of worker processes via the socket activation logic was so that the risky processing can be sandboxed individually, and we can reasonably safely process ELF stuff. The coredump processing tools are thus already pretty unprivileged/sandboxed and it's OK if they die abnormally (we'll lose the coredump then, but that's OK). But now you want to split this up a second time? so that the locked down per-coredump handler process forks itslef and does part of what it does in a subprocess again?

Reply:
there's also this issue that we shouldn't do long running stuff in a fork() child without execve() first. glibc made clear that a good chunk of their APIs are not ready for that, it's an OOM trap and so on. i.e. we should ensure that our fork() children are always very short-lived. Here however we do expensive operations in the child, hence there really should be an execve() in between.

We didn't use to be strict on that, but we should fix our codebase regarding this. For example, we discussed plans a while back to change PID1's main process forking logic to operate quite differently from now: instead of doing its myriad of initializations in the child right after fork(), first serialize what is to be done to JSON and then do an execve() to an "executor" tool of our own, that then reads the JSON and applies it. That's important because glibc people are not actually OK with how we use NSS for User=/Group= and suchlike (NSS is one of those APIs one isn#t supposed to call in a fokred child before execve()), but more importantly PAM (which is a total mess right now: we leave the PAM session hook process around for a service's entire lifetime).

TLDR: for new code we should not introduce new long running processes that are fork()ed but not execve()ed. 

a possible fix here is to just exec /proc/self/exe here, and pass the few fds we need over, and then dispatch things from there.

Reply:
yes - because it was shown that libdw is very crashy when handling garbled data (@keszybz says there are bug reports from Fedora with v249), and what we are looking for here is essentially optional data. it's good if we can get it, but if not, that's fine - but we shouldn't lose the core, the log in the journal etc because of this.

Reply:
> TLDR: for new code we should not introduce new long running processes that are fork()ed but not execve()ed.

I get what you are saying, but is this really a long running one? It doesn't seem fundamentally different from doing a dissect on a big binary image to me - it's just looking at the headers of a single file

Reply:
extracting stacktraces from huge coredumps is really slow actually. i.e. firefox coredump processing can take ages since the binaries are so large and with debug symbols and stuff its apparently really a lot of work to get to the stacktraces

Reply:
(and new elfutils could even talk to debuginfod, if people enable that, making it even slower in the not distant future)

Reply:
I get that, but it's not long running in the sense that it does many different things, services others, calls into glibc APIs or any of that - we are running network-sandboxed too, so no debuginfod (and it's one of the reasons we pushed for the metadata to be directly in the file on https://fedoraproject.org/wiki/Changes/Package_information_on_ELF_objects#Why_not_just_use_debuginfod.3F ). Yeah for big files it might take a while to iterate over and compress, but it's all strictly upper bounded and limited. I mean from what I hear from @keszybz the crash issue is pretty serious, so it's a real actual problem that we have right now, that this solves.

Reply:
so, i get that there might be value in still generating a journal record even if processing the stacktrace fails, hence I am OK to fork off a child for this. But please add a comment explaining this.

And yes, debuginfod by default is not in the cards, but I'd allow it for the future as an option if people send us a patch. 

Still, if you do more than a bunch of preparatory syscalls, please add an execve() in there. It shouldn't require much glue, there's very little data you want to pass over: just the fds basically: hence doing execve() on /proc/self/exe with the fds passed over should be easy, and then redirecting things back from main() should be really easy.

(why go via /proc/self/exe btw, instead of a `#define`d path of the coreudp binary btw? simply to avoid version incompat for the data passed over)

Reply:
> Still, if you do more than a bunch of preparatory syscalls, please add an execve() in there. It shouldn't require much glue, there's very little data you want to pass over: just the fds basically: hence doing execve() on /proc/self/exe with the fds passed over should be easy, and then redirecting things back from main() should be really easy.

It's called from multiple programs, one of which already receives FDs on input (systemd-coredump), so it's not really trivial or clean. But again, what is the point? There's no system calls apart from reading from a FD and writing to the local pipes. No interaction with gblic, no interaction with the system in any way, no calling into other libraries apart from libelf and libdw, just looping over local memory. Stuff like systemd-dissect do so much more in the forked processes - open images, decrypt volumes, talk with the kernel, etc.

Reply:
To comment just on the processing in a separate process: I'm pretty sure we need to do that. Right now we have the situation where systemd-coredump crashes in the backtrace generation code quite often. Not only we lose the original report, but also (since this was coredump itself crashing), we don't get a proper report for the coredump crash. So the original reproducer is gone (the first corefile) and we don't get good info for the second crash, and we can't improve this. If we do the processing in a seperate process, if that crashes, we will report the original crash, including the core file, and the second crash too. This makes it much more likely that a report with enough data to improve systemd-coredump will be made.